### ISSUE TITLE
Jetty-12 `Content.Provider` Producer?

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#7403]
**Jetty version**
12

**Question**
After the recent conversations regarding websocket and IteratingCallback I wanted to make sure that Jetty-12 has a good processor abstraction that handles both demand and release correctly.
But I'm currently tying myself in knots trying to come up with right abstract for how a truly general processor should work (ie one that can both silently consume and spontaneously produce content(s)).

So currently the `Content` signature is:

```java
public interface Content
{
    ByteBuffer getByteBuffer();
    boolean isSpecial();
    boolean isLast()
    void release();
}
```

and we have a Provider abstraction like:
```java
    interface Provider
    {
        Content readContent();
        void demandContent(Runnable onContentAvailable);
    }
```

So the challenge is to make a utility processor class with a signature like:
```java

public abstract class ContentProcessor implements Content.Provider
{
    public ContentProcessor(Content.Provider provider) { /* ... */ }

    public Content readContent() { /* ... */ }
    public  void demandContent(Runnable onContentAvailable) { /* ... */ }

    protected abstract Content produce(Content content);
}
```

The idea being that the processor will call the provider for demand/content as needed and then call `produce(Content content)` to transform that content somehow:
 + content may be passed straight through, in which case it is the responsibility of the ultimate consumer to release it.
 + content may be translated on a 1:1 basis to new content, in which case the processor must release the passed content and the consumer must release the produced content.
 + content may be consumed without producing any content.  It may or may not be released when it is consumed and it's remaining count may or may not be none zero.
 + we need to allow consumed content to produce multiple contents.
 + perhaps we should allows a processor to spontaneously produce content without any input content????

So the big question I have, is what exactly is the contract with the `Content produce(Content content)` method. Specifically how does it indicate that it has finished with the passed content and the processor should try to read more content from the provider?

For the purposes of the following discussion, I am going to user the name "input" for content read from the provider and passed to the `produce` method.  I will use the name "output" for content returned from the `produce` method. Yeah I know, not strictly correct but will do for the purposes of discussion.

**Option 1**

The same input content is passed repeatedly to the `produce` method until such time as it returns null.  So consider a producer that:
 + splits string content on white space into new content
 + consumes any strings "NOOP"
 + Copies any other upper case content into a new lower case content
 + passes all other content straight through

So if the following queue was available:
> "one"
> "TWO"
> "three four"
> "NOOP"
> "five NOOP six"
there would be the following sequence of calls and returns (note that I'm not showing any consumptions of bytes in the content)
```
 provider.readContent() -> "one"
 produce("one") -> "one"
 produce("one") -> null
 provider.readContent() -> "TWO"
 produce("TWO") -> "two"
 produce("TWO") -> null
 provider.readContent() -> "three four"
 produce("three four") -> "three"
 produce("three four") -> "four"
 produce("three four") -> null
 provider.readContent() -> "NOOP"
 produce("NOOP") -> null
 provider.readContent() -> "five NOOP six"
 produce("five NOOP six") -> "five"
 produce("five NOOP six") -> "six"
 produce("five NOOP six") -> null
```

PROS:
 + processor doesn't need to remember the input as it keeps getting passed in until it is finished

CONS:
 + two calls to `produce` for every content in a 1:1 scenario
 + if the processor releases the input content but returns non null, it is *wrong* to pass a released content to the processor on subsequent calls as the bytebuffer may be reused and contain anything!  So processors would need to delay releasing until they return null.
 + If the processor uses release to signal input demand, then it may be forced to hold onto large content longer than is optimal to signal back pressure.

**Option 2**
The same as Option 1 but input is represented until such time as null is return OR it is released.   So for the same example producer and same input:

```
 provider.readContent() -> "one"
 produce("one") -> "one"
 produce("one") -> null
 provider.readContent() -> "TWO"
 produce("TWO") -> "two"  && "TWO".release()
 provider.readContent() -> "three four"
 produce("three four") -> "three"
 produce("three four") -> "four" && "three four".release()
 provider.readContent() -> "NOOP"
 produce("NOOP") -> null
 provider.readContent() -> "five NOOP six"
 produce("five NOOP six") -> "five"
 produce("five NOOP six") -> "six" && "five NOOP six".release()
```

PROS:
 + If the processor does the releasing, you can avoid lots of null returning calls and get close to one call to `produce` for every content in a 1:1 scenario

CONS:
 + It's a little subtle and complex.
 + input content needs to be wrapped to intercept release

**Option 3**
A content is passed to the processor once and only once. Instead, it is called with null to see if anything can be produced without more input.   So for the same example, the calling sequence would be:
```
 produce(null) -> null
 provider.readContent() -> "one"
 produce("one") -> "one"
 produce(null) -> null
 provider.readContent() -> "TWO"
 produce("TWO") -> "two"
 produce(null) -> null
 provider.readContent() -> "three four"
 produce("three four") -> "three"
 produce(null) -> "four"
 produce(null) -> null
 provider.readContent() -> "NOOP"
 produce("NOOP") -> null
 provider.readContent() -> "five NOOP six"
 produce("five NOOP six") -> "five"
 produce(null) -> "six"
 produce(null) -> null
```

PROS:
 + the contract is simple to describe.
 + no issues with releasing the input (can be done at anytime by either processor or consumer)
 + the processor can produce output even before any input has arrived.
 + It is the only option I've found that has a clear signals for "input release" and "input demand" that does not allow processor abuse.

CONS:
 + lots of `produce(null)` calls
 + the processor has to keep a reference to the input if it wants to use it multiple times 

**Option 4**
Have a produce that returns a boolean like `boolean produce(Content input, Consumer<Content> output)`, that returns true when the processor is finished with the input.
PROS:
 + The boolean makes it a lot clearer when the producer no longer wants to deal with the input.

CONS:
 + We still have the issue of a producer that returns false, but has released the input... although that could be fixed by only passing an input once and then passing null until a true return
 + the output consumer can be called multiple times for each produce call, which can violate demand.

**Option 5**
The producer could be given the option to control it's own input demand like `Content produce(Content input, Runnable demandInput)`
PROS:
 + probably will work

CONS:
 + complicates the producer so there is significant risk of it not working

**Option 6**
Somehow use bytes remaining????

CONS:
 + Doesn't well represent a fully consumed input  that can still produce multiple outputs.  We don't want to request more input until it is truly finished.
 + Doesn't allows content to be held by processor nor consumer.

**Discussion**
I'm not exactly sure which one I like (or which I dislike least)..... but probably going to try to implement **option 3** to test.       However, this now put's yet another requirement on our consideration for better APIs/contracts in jetty-12.   Not only does anything we come up with have to work with all the jetty-server unit-tests that are currently passing, but I think we need to show that it has a good processor behavior as well.      The output side is not so bad as that is just traditional write with `Callback` and neither demand nor release are relevant.

thoughts?

**Comment by gregw** [comment]
@sbordet @lorban @lachlan-roberts thoughts?

**Comment by lorban** [comment]
So you're looking at creating a new generic API which would allow multiple implementations of this method:
```java
protected abstract Content produce(Content content);
```
I'm really wondering if that's worth it. The `Provider` interface already has a well-defined contract w.r.t the lifecycle of the `Content` object returned by the `readContent()` method: it's up to the caller to manage it.

With that in mind, I think it would be fairly simple to implement a `GzipProvider` that would override any other provider by overriding the `readContent()` method and following its contract, both as a provider (implementing it) and a consumer (calling the delegate). Your string-splitting-with-NOOPs example would be fairly easy to implement too.

My point is that I'm not sure a generic mechanism is worth it and IMHO keeping the bridging code between the wrapping `Provider` and the wrapped one private to the wrapping implementation would be best. As long as the existing `readContent()` contract is followed, the wrapper is free to keep wrapped `Content` references around, succeed them when it sees fit and return new `Content` instances as it wants.

The only benefit to that new API is if for some reason you'd like to mix returning new `Content` instances and original ones but that looks like an anti-pattern to me.

**Comment by sbordet** [comment]
I also don't like much `Content produce(Content content)` given that it needs to rely on something else to demand for more content, etc. -- it's not self contained.
Yes perhaps you can rely on the whole utility class, but then I think there would be a confusion about "do I implement produce() or do I call readContent()?"
So a processor of content is nothing more than:

```java
abstract class Processor implement Provider {
  Processor(Provider previous) {
    ...
  }
  ...
}
```
where the `previous` passed to the constructor is the upstream stage, and an instance of `Processor` is the current stage.

Compare this to reactive-streams: it has a `Subscriber` class with `onSubscribe(), onNext(T), onComplete(), onError()`.

All of these functionalities (aside `onSubscribe()`, replaced by the constructor) have been merged into `readContent()`, where `Content` reports whether it's the next, the last, or an error.

In reactive-streams, `Subscription` had the demand primitive, but here again we coalesced it into `Provider` because we use a pull model (`onContentAvailable()`+`readContent()`) rather than a push model (`onNext(T)`).

So in reactive-streams a Processor was a Publisher+Subscriber, but here we don't have a publisher abstraction because it's implicit in `Provider` (because an app is given a first `Provider` onto which it demands, exactly like a reactive-streams app is given a first `Publisher` into which register a `Subscriber`); furthermore, `Provider` is also the `Subscription` (because it has `demandContent()`) and also the `Subscriber` (because it has `Content readContent()`).

One things that we discussed at the time with the reactive-streams folks is how to propagate errors.
Upstream to downstream is easy: upon demand, `onContentAvailable()` is called, and upstream produces an error `Content` when `readContent()` is called.
Downstream to upstream is not done: we just stop demanding, but we may need to propagate the error upstream to clear up resources.

Lastly, while reactive-streams is generic (works with `Content` but also with `Frame`s, etc.), this mechanism seems quite peculiar to just one use case (reading `ByteBuffer` content), so a `Processor` abstract class seems "meh" to me -- perhaps it's useful, perhaps not, but not a big deal.

I'm more interested in analyzing upstream error propagation for upstream resource release.

**Comment by gregw** [comment]
@sbordet yeah this is indeed really just a Processor in the style of reactive streams.     So it could indeed be modelled as an API and a utility class:
```java
interface Content
{
    // ...

    abstract class Processor implements Provider
    {
        private final Provider _provider;
        protected Processor(Provider provider)
        {
            _provider = provider;
        }
    }
}
```

Which is really just a wrapper by another name.
The `ContentProcessor` class is just a utility class to solve the problem generally... mostly at this stage to prove that it can be done to test the `Provider` contract.   I'll make that change in my branch as I think it is good to separate form from utility implementation.

However, I don't see this as semantically that different to Reactive Streams.  The main difference is that we pull content/errors with `readContent` rather than have it delivered via `onNext` and `onError`.     I thought we had decided that is a good thing, and I still like it (plus it kind of matches our EndPoint contract of fillInterested and fill).     The big difference to Reactive Streams, is that we have a release semantic also built into the API which complicates things, so when you say "perhaps not, but not a big deal", how sure of that are you?   I think you need to write some actual code in jetty-12 to either confirm that and/or to come up with better abstractions

Ditto @lorban   When you say: "I think it would be fairly simple to implement a `GzipProvider`" how sure of that are you?     It would also need to solve all the same issues of dealing with demand and release, just as does my `ContentProcessor`.     It's not exactly difficult, once a solution is designed, but it has been tricky to design in such a way that locks are not being grabbed and released too much.

With regards to "you'd like to mix returning new Content instances and original ones but that looks like an anti-pattern to me."   I agree that in any once Processor it is probably an anti pattern to mix too much.    But then there are use-cases from the websocket world, where they have fragmentation extensions - they pass on small frames unchanged, but will fragment big frames into new smaller frames.       Again, I'm not really advocating that a processor mixes modes unless there is a good reason, it is really just that I want to test that the API works either if the processor passes content on, or if it makes new content.  A gzip processor would always make new content, so perhaps it is simpler in that regard.

The proof is in working code.  I really think @lorban  and/or @sbordet  should write the GzipHandler in jetty-12 to get your heads into the code.   You both need to engage with it and actually write working code in that branch, because currently I'm sensing lots of mild dissapproval, but not many alternative approaches.   Time to get your hands dirty... even if it is just to rewrite everything to an alternative!

I'll do a PR of my branch later today so you can see where the actual implementation is at.

**Comment by sbordet** [comment]
> However, I don't see this as semantically that different to Reactive Streams.

Not that different, but:

* ReactiveStreams is generic on type `T`; here works only with ByteBuffers.
* ReactiveStreams may have multiple `Subscriber`s for one `Publisher`, each with its own demand; here we only have one thing.
* ReactiveStreams does not have a `release` primitive; here we do.
* ReactiveStreams has a partial upstream `cancel()` primitive; here we don't (at all).

To clarify "not a big deal" refers to implementing an abstract utility `Processor` class.
Sure it'll be needed for a couple use cases, but it's not a big deal to implement.
The exact definition of the `Provider` interface is more important than implementing a utility class.

> The proof is in working code.

I prefer well thought out interfaces first, because I can get crappy working code very easily.

> currently I'm sensing lots of mild dissapproval, but not many alternative approaches

I don't think we're discussing the `Provider` interface here.
I still don't like much `demandContent(Runnable)`, but turns out that `EndPoint.fillInterested(Callback)` has the same semantic, so I'm relaxing a bit 😄 

However, don't mix `Content.Provider` with `EndPoint` or WebSocket, because there is a big difference in the threading model.
First of all `Content.Provider` cannot be used with WebSocket frames. I also won't use it with `EndPoint` because I don't want to create `Content` instances every time.
So, similar ideas, but different abstractions.
Furthermore, WebSocket extensions and `EndPoint` have a different threading model.
The big complication of implementing a `Processor` is that `demandContent()` and `readContent()` can be called at any time by any thread, because it's unknown application code that invokes those methods (see how many locks and state you need to hold).
For WebSocket extensions and `EndPoint` we completely control who's calling from what thread, so implementation are typically much simpler.

There is no disapproval, just discussing.
I have done the exercise of implementing `Processor` and I agree with you that a utility class is probably needed, because `demandContent()` and `readContent()` can be called at any time so locking is necessary and it's easy to get it wrong.
The moment you start implementing `demandContent()` you have the problems "what if it's called concurrently", "what if it's called multiple times", "how does it coordinates with a concurrent call to `readContent()`", "how does it coordinate with an upstream demand", and so on. 
So I started sprinkling `AtomicReferences` around until I realized I have to use a `Lock`.
Point taken about "get your hands dirty" but there is more to fix in the design.

**Comment by lorban** [comment]
Something that caught my attention and upon which we seem to disagree is the threading model of the `demandContent()` and `readContent()` methods. While I do agree that  `demandContent()` must be thread-safe, I fail to see what a concurrent `readContent()` would bring, and how it would work.

**Comment by gregw** [comment]
@lorban I agree that there is little use-case for a call to `readContent()` in parallel with `demandContent()`, but we should try to act reasonably.   My impl of `readContent()` just returns null if it sees a concurrent call to either `readContent()` or `demandContent`, which I think is a reasonable failsafe and kind of validates the API.

@sbordet 
> I prefer well thought out interfaces first, because I can get crappy working code very easily.

I kind of think that I'm at the stage of testing in the API is well thought out or not by seeing if we can get some non-crappy working code.  Hence I'm trying out multiple things to see if the API is really OK or not.  It's good that it sounds like you've done some code against it and that you are "relaxing a bit", as I'm kind of the same.   I don't like aspects of the API, but when I try to fix them, I'm still finding what I've got the best compromise I can think of.       So we need to keep pushing this to either all come to grudging acceptance of the proposal  OR to really decide we don't like it and find something better.

> Point taken about "get your hands dirty" but there is more to fix in the design.

I definitely agree that the design is far from set in stone, and more thinking and dirty hands are needed to both test current design, but also to try-out any alternatives we can think of.  
 So there are lots of dirty-hand tests that can be done to test the current code base: GzipHandler, Idle Timeout, Upgrade, RequestLogging, etc.       Or changing the API/design and trying to pass all the current tests in jetty-server.      Just yell if there is something I can do to free up more of your time to play in this code base, as we really need it to be common-ownership, in that we all accept it as the best way forward.

**Comment by gregw** [comment]
closing, as we have progressed and any API changes will come with subsequent review and new issues.

**Comment by joakime** [comment]
PR #7404 merged.
